File: python/src/train/models/din.py
"""Deep Interest Network (DIN, arXiv:1706.06978)."""

from typing import TYPE_CHECKING

import torch
import torch.nn as nn

if TYPE_CHECKING:
    from ..core.config import PoolingMode

from ..core.model_output import ModelExecution, ModelOutput
from ..core.output_contract import NormalizedOutputContract
from ..layers.embedding import FeatureEmbeddings, FeatureTensorMap, FeatureTuple
from ..layers.mlp import Mlp
from ..layers.towers import Activation
from .output_head import OutputHead


class DIN(nn.Module):
    """Deep Interest Network.

    Shared item embedding for behavior sequence + candidate ad.
    Local activation unit computes adaptive, unnormalized user interest weights.
    """

    def __init__(
        self,
        features: list[FeatureTuple],
        item_vocab_size: int = 10000,
        embed_dim: int = 16,
        activation_hidden_dims: list[int] | None = None,
        mlp_hidden_dims: list[int] | None = None,
        behavior_feature: str = "hist_item_ids",
        candidate_feature: str = "item_id",
        pooling_map: dict[str, PoolingMode] | None = None,
        total_dim: int | None = None,
        output_contract: NormalizedOutputContract | None = None,
    ) -> None:
        super().__init__()
        self.embed_dim = embed_dim
        self.behavior_feature = behavior_feature
        self.candidate_feature = candidate_feature

        self.item_embedding = nn.Embedding(item_vocab_size, embed_dim)

        other_features = [
            feature
            for feature in features
            if feature[0] not in {behavior_feature, candidate_feature}
        ]
        self.embeddings = FeatureEmbeddings(other_features, pooling_map) if other_features else None
        other_dim = self.embeddings.total_dim if self.embeddings is not None else 0

        self.activation_unit = Mlp(
            4 * embed_dim,
            activation_hidden_dims or [],
            1,
            Activation.RELU,
        )

        mlp_input_dim = embed_dim + embed_dim + other_dim
        self.mlp = Mlp(
            mlp_input_dim,
            mlp_hidden_dims or [],
            1,
            Activation.RELU,
        )

        self.output_head = (
            OutputHead(output_contract, {"shared": 1}) if output_contract is not None else None
        )

    def forward(self, x_inputs: FeatureTensorMap) -> ModelOutput:
        if self.output_head is not None:
            return self.forward_execution(x_inputs).outputs
        return ModelOutput.binary_logits({"pred": self._shared(x_inputs)})

    def forward_execution(self, x_inputs: FeatureTensorMap) -> ModelExecution:
        if self.output_head is not None:
            return self.output_head({"shared": self._shared(x_inputs)})
        outputs = self.forward(x_inputs)
        return ModelExecution(nodes=outputs, outputs=outputs)

    def _shared(self, x_inputs: FeatureTensorMap) -> torch.Tensor:
        behavior_idx = x_inputs[self.behavior_feature]
        candidate_idx = x_inputs[self.candidate_feature]

        if behavior_idx.dim() == 1:
            behavior_idx = behavior_idx.unsqueeze(1)
        behavior_embs = self.item_embedding(behavior_idx)
        candidate_emb = self.item_embedding(candidate_idx)

        batch, seq_len, embed_dim = behavior_embs.shape

        candidate_emb_3d = candidate_emb.unsqueeze(1).expand(batch, seq_len, embed_dim)
        prod = behavior_embs * candidate_emb_3d
        diff = behavior_embs - candidate_emb_3d

        au_input = torch.cat([behavior_embs, candidate_emb_3d, prod, diff], dim=2)
        au_input_flat = au_input.reshape(batch * seq_len, 4 * embed_dim)
        attn_weights = self.activation_unit(au_input_flat).reshape(batch, seq_len, 1)

        interest_emb = (behavior_embs * attn_weights).sum(dim=1)

        if self.embeddings is not None:
            other_emb = self.embeddings(x_inputs)
            combined = torch.cat([interest_emb, candidate_emb, other_emb], dim=1)
        else:
            combined = torch.cat([interest_emb, candidate_emb], dim=1)
        return self.mlp(combined)

File: src/models/din.rs
//! Deep Interest Network (DIN, arXiv:1706.06978).
//!
//! DIN uses a local activation unit to adaptively learn user interest
//! representation from behavior sequences with respect to a candidate ad.
//! Shared item embedding table for both behavior items and candidate ad.

use super::output_contract::OutputContract;
use super::output_head::OutputHead;
use super::{Model, ModelExecution, ModelOutput};
use crate::layers::embedding::{FeatureEmbeddings, FeatureSpec};
use crate::layers::mlp::Mlp;
use crate::layers::towers::Activation;
use candle_core::{Result, Tensor};
use candle_nn::{embedding, Module, VarBuilder};
use std::collections::HashMap;

/// Deep Interest Network model.
pub struct DIN {
    item_embedding: candle_nn::Embedding,
    embeddings: Option<FeatureEmbeddings>,
    activation_unit: Mlp,
    mlp: Mlp,
    output_head: Option<OutputHead>,
    embed_dim: usize,
    behavior_feature: String,
    candidate_feature: String,
}

impl DIN {
    /// Construct DIN with shared item embedding + activation unit + MLP.
    pub fn new(
        vb: VarBuilder,
        features: &[FeatureSpec],
        item_vocab_size: usize,
        embed_dim: usize,
        activation_hidden_dims: &[usize],
        mlp_hidden_dims: &[usize],
        behavior_feature: &str,
        candidate_feature: &str,
    ) -> Result<Self> {
        let item_embedding = embedding(item_vocab_size, embed_dim, vb.pp("item_embedding"))?;

        let other_features: Vec<FeatureSpec> = features
            .iter()
            .filter(|feature| feature.name != behavior_feature && feature.name != candidate_feature)
            .cloned()
            .collect();
        let embeddings = if other_features.is_empty() {
            None
        } else {
            Some(FeatureEmbeddings::new(
                vb.pp("embeddings"),
                &other_features,
            )?)
        };
        let other_dim = embeddings
            .as_ref()
            .map_or(0, |embeddings| embeddings.total_dim);

        let activation_unit = Mlp::new(
            vb.pp("activation_unit"),
            4 * embed_dim,
            activation_hidden_dims,
            1,
            Activation::Relu,
        )?;

        let mlp = Mlp::new(
            vb.pp("mlp"),
            embed_dim + embed_dim + other_dim,
            mlp_hidden_dims,
            1,
            Activation::Relu,
        )?;

        Ok(Self {
            item_embedding,
            embeddings,
            activation_unit,
            mlp,
            output_head: None,
            embed_dim,
            behavior_feature: behavior_feature.to_string(),
            candidate_feature: candidate_feature.to_string(),
        })
    }

    /// Construct DIN with output contract.
    pub fn with_output_contract(
        vb: VarBuilder,
        features: &[FeatureSpec],
        item_vocab_size: usize,
        embed_dim: usize,
        activation_hidden_dims: &[usize],
        mlp_hidden_dims: &[usize],
        behavior_feature: &str,
        candidate_feature: &str,
        contract: &OutputContract,
    ) -> Result<Self> {
        let mut model = Self::new(
            vb.clone(),
            features,
            item_vocab_size,
            embed_dim,
            activation_hidden_dims,
            mlp_hidden_dims,
            behavior_feature,
            candidate_feature,
        )?;
        let representation_dims = HashMap::from([("shared".to_string(), 1)]);
        model.output_head = Some(OutputHead::new(
            contract,
            &representation_dims,
            vb.pp("output_head"),
        )?);
        Ok(model)
    }

    fn shared(&self, x_inputs: &HashMap<String, Tensor>) -> Result<Tensor> {
        let behavior_indices = x_inputs.get(&self.behavior_feature).ok_or_else(|| {
            candle_core::Error::Msg(format!(
                "Missing behavior feature '{}'",
                self.behavior_feature
            ))
        })?;
        let candidate_indices = x_inputs.get(&self.candidate_feature).ok_or_else(|| {
            candle_core::Error::Msg(format!(
                "Missing candidate feature '{}'",
                self.candidate_feature
            ))
        })?;

        let behavior_indices = if behavior_indices.rank() == 1 {
            behavior_indices.unsqueeze(1)?
        } else {
            behavior_indices.clone()
        };
        let behavior_embs = self.item_embedding.forward(&behavior_indices)?;
        let candidate_emb = self.item_embedding.forward(candidate_indices)?;

        let seq_len = behavior_embs.dim(1)?;
        let embed_dim = self.embed_dim;
        let batch = behavior_embs.dim(0)?;

        let candidate_emb_3d = candidate_emb
            .unsqueeze(1)?
            .broadcast_as(behavior_embs.shape())?;
        let prod = behavior_embs.mul(&candidate_emb_3d)?;
        let diff = behavior_embs.sub(&candidate_emb_3d)?;

        let au_input = Tensor::cat(&[&behavior_embs, &candidate_emb_3d, &prod, &diff], 2)?;
        let au_input_flat = au_input.reshape((batch * seq_len, 4 * embed_dim))?;
        let attn_weights = self
            .activation_unit
            .forward(&au_input_flat)?
            .reshape((batch, seq_len, 1))?;

        let interest_emb = behavior_embs
            .mul(&attn_weights.broadcast_as(behavior_embs.shape())?)?
            .sum(1)?;

        match &self.embeddings {
            Some(embeddings) => {
                let other_emb = embeddings.forward(x_inputs)?;
                let combined = Tensor::cat(&[&interest_emb, &candidate_emb, &other_emb], 1)?;
                self.mlp.forward(&combined)
            }
            None => {
                let combined = Tensor::cat(&[&interest_emb, &candidate_emb], 1)?;
                self.mlp.forward(&combined)
            }
        }
    }
}

impl Model for DIN {
    fn forward(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelOutput> {
        if self.output_head.is_some() {
            return Ok(self.forward_execution(x_inputs)?.outputs);
        }
        let logits = self.shared(x_inputs)?;
        let mut outputs = ModelOutput::new();
        outputs.insert_binary_logit("pred", logits);
        Ok(outputs)
    }

    fn forward_execution(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelExecution> {
        if let Some(head) = &self.output_head {
            return head.forward(&HashMap::from([(
                "shared".to_string(),
                self.shared(x_inputs)?,
            )]));
        }
        let outputs = self.forward(x_inputs)?;
        Ok(ModelExecution::new(outputs.clone(), outputs))
    }
}
